lla_to_ecef.cpp

```diff
@@ -0,0 +1,390 @@
+
+    // Exit if point of interest is out of range
+    if ((ts_is_before(point_of_interest, lla_data[0].t))
+            || ts_is_before(lla_data.back().t, point_of_interest) ) {
+        return indices;
+    }
+    
+    int idx = 0;
+    for (const auto &entry : lla_data) {
+        if (ts_is_equal(entry.t, point_of_interest)) {
+            indices[1] = idx;
+            break;
+        } else if (ts_is_before(entry.t, point_of_interest)) {
+            idx++;
+            continue;
+        } else {
+            indices[0] = idx - 1;
+            indices[2] = idx;
+        }
+    }
+
+    return indices;
+}
+
+/**
+ * lla_to_ecef_pos
+ * @brief Converts a position in LLA coordinates to ECEF coordinates.
+ *
+ * Uses closed formulas to transform from LLA to ECEF coordinates:
+ *
+ *    x_ecef = (r_curve + h) * cos(lat) * cos(long)
+ *    y_ecef = (r_curve + h) * cos(lat) * sin(long)
+ *    z_ecef = ( ((b * b * r_curve) / (a * a)) + h) * cos(lat) * cos(long)
+ *
+ * where a and b are the semi-major and semi-minor axes of the reference
+ * ellipsoid in meters, h is the height above the reference ellipsoid in meters,
+ * and N is the radius of curvature, defined as:
+ *
+ *    r_curve = a / sqrt[1 - ( (e * e) * (sin(lat) * sin(lat)) )]
+ *
+ * For more information see Datum Transformation of GPS Positions, at:
+ * https://community.ptc.com/sejnu66972/attachments/sejnu66972/PTCMathcad/57177/2/Datum%20Transformations.pdf
+ *
+ * @param pos_lla a PositionLLA struct containing:
+ *                  - time as timespec struct, since Unix epoch
+ *                  - latitude [degrees]
+ *                  - longitude [degrees]
+ *                  - altitude data [meters]
+ * @param a       reference ellipsoid semi-major axis [meters]
+ * @param b       reference ellipsoid semi-minor axis [meters]
+ * @param e       reference ellipsoid eccentricity [meters]
+ *
+ * @return a PositionVelocityECEF struct with the following:
+ *                  - same as LLA position: time as timespec struct, since Unix epoch
+ *                  - x, y, and z ECEF position fields as calculated, in meters
+ *                  - x, y, and z ECEF velocities are left uninitialized
+ */
+PositionVelocityECEF lla_to_ecef_pos(
+    PositionLLA pos_lla, double a, double b, double e) {
+    double r_curve = get_rad_of_curvature(a, e, pos_lla.latitude);
+
+    double c_lat = std::cos(pos_lla.latitude * DEG_TO_RAD);
+    double s_lat = std::sin(pos_lla.latitude * DEG_TO_RAD);
+    double c_lon = std::cos(pos_lla.longitude * DEG_TO_RAD);
+    double s_lon = std::sin(pos_lla.longitude * DEG_TO_RAD);
+
+    double x_ecef = (r_curve + pos_lla.altitude) * c_lat * c_lon;
+    double y_ecef = (r_curve + pos_lla.altitude) * c_lat * s_lon;
+    double z_ecef = ((std::pow(b, 2) * r_curve / std::pow(a, 2))
+                        + pos_lla.altitude) * s_lat;
+
+    return PositionVelocityECEF {pos_lla.t, x_ecef, y_ecef, z_ecef};
```

Comment:
any concern about declaring the variables here

doubles should be implicitly copied but im a control freak and an embedded guy so I don't assume any rules of memory